src/validation.cpp

```diff
@@ -1295,6 +1312,15 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::
             results.emplace(ws.m_ptx->GetWitnessHash(), MempoolAcceptResult::Failure(ws.m_state));
             return PackageMempoolAcceptResult(package_state, std::move(results));
         }
+
+        // Individual modified feerate exceeded caller-deemed "sane" max; abort
```

Comment:
My understanding of package submission is that a child with unconfirmed parents is bundled together as a package for incentive compatibility. Should this instead check the package fee rate against `maxfeerate`?

If I understand correctly, currently, we are checking the package transaction fee rate individually against `maxfeerate`. The child transactions might be greater than the `maxfeerate`, but as a package might not be. In this case, we might reject the child transactions and accept the subpackage, which might not be incentive compatible at the current fee estimates.

For example, suppose we want to fee bump tx A, with the arrangement below. `maxfeerate` is set to 50 s/vb.

```md
               All txs are 10vb
               first value is absolute fee
               second value is ancestor score
               ┌───────────┐
               │  A (10s)  │
               │ 1 s/vb    │
               └───────────┘
                     │
                     ▼
               ┌───────────┐
               │  B (90s)  │
               │ 5 s/vb    │
               └───────────┘
                     │
                     ▼
               ┌────────────┐
               │  C (1000s) │
               │ 36.6 s/vb  │
               └────────────┘

```
If `maxfeerate` is 50s/vb we will accept A-B and reject C because it's ancestor score is 100s/vb.
But if we are evaluating maxfeerate against package fee rate we will accept the whole package A-B-C

Reply:
Ignoring topo restrictions in `AcceptPackage` since `C` should have `A` as a direct ancestor, 

> The child transactions might be greater than the maxfeerate, but as a package might not be. In this case, we might reject the child transactions and accept the subpackage, which might not be incentive compatible at the current fee estimates.

This is prior to any relay, of course restricting feerates is incentive compatible once published on the open network!

> If maxfeerate is 50s/vb we will accept A-B and reject C because it's ancestor score is 100s/vb.
But if we are evaluating maxfeerate against package fee rate we will accept the whole package A-B-C

Assuming you meant C's *feerate* is 100 sat/vbyte: If mempool minfee is 5 sat/vbyte, A+B would be accepted(as they'd be evaluated in their own subpackage already), then C evaluated on its own `AcceptSingleTransaction`, and rejected.

In a post-cluster mempool world, this would likely be reduced to evaluating chunks only, and so you'd look at A+B+C only, and reject the whole thing.

Reply:
>Assuming you meant C's feerate is 100 sat/vbyte

Yes C fee rated is 100s/vb, edited thanks.

>This is prior to any relay, of course restricting feerates is incentive compatible once published on the open network

Do you mean incentive incompatible here? because in the given example if C is rejected A will be  added to the mempool with mining score of 5s/vb which is lower than whats intended.

Reply:
> Do you mean incentive incompatible here? because in the given example if C is rejected (A,B) will be added to the mempool with mining score of 5s/vb which is lower than whats intended.

Comment
err right
 
src/rpc/client.cpp
Comment on lines +130 to +131
    { "submitpackage", 1, "maxfeerate" },
    { "submitpackage", 2, "maxburnamount" },
Member
@ismaelsadeeq ismaelsadeeq 2 days ago • 
Should we just enable passing maxfeerate and maxburnamount as a config variables since it is used three RPC's now sendrawtransaction, testmempoolaccept and submitpackage, downstream might prefer this?

Member
Author
@instagibbs instagibbs 2 days ago
startup config? Maybe? Could open an issue

Member
@ismaelsadeeq ismaelsadeeq 3 hours ago
Done in #29217

@instagibbs	Reply...
src/rpc/mempool.cpp
Comment on lines +889 to +890
            const CAmount max_burn_amount = request.params[2].isNull() ? 0 : AmountFromValue(request.params[2]);

Member
@ismaelsadeeq ismaelsadeeq 2 days ago • 
We did not document this anywhere but the default behavior if maxburnamount is not passed we are rejecting the package if there is any transactions with unspendable output greater than 0.
Maybe indicate it in the RPCHelpMan of maxburnamount?

Its also not indicated for sendrawtransaction.

Member
Author
@instagibbs instagibbs 2 days ago
added a mention of the default being 0

@instagibbs	Reply...
test/functional/rpc_packages.py
Outdated
        assert_raises_rpc_error(-25, "Unspendable output exceeds maximum configured by user", node.submitpackage, chain_hex, 0, 0.1)
        assert_equal(node.getrawmempool(), [])

        # Turn off restrictions for both and send it; parent gets through as own subpackage
Member
@ismaelsadeeq ismaelsadeeq 2 days ago
To turn off for maxfeerate you should pass 0 fee rate, you cant turn off the restriction of maxburnamount
nit: maybe reword to relax?

Suggested change
        # Turn off restrictions for both and send it; parent gets through as own subpackage
# Relax the restrictions and send it; parent gets through as own subpackage
Member
Author
@instagibbs instagibbs 2 days ago
taken

@instagibbs	Reply...
@instagibbs
RPC: Add maxfeerate and maxburnamount args to submitpackage 
bf6fbea
@instagibbs instagibbs force-pushed the 2023-11-submitpackage-max-fee-burn branch from e650eed to bf6fbea 
2 days ago
@ismaelsadeeq ismaelsadeeq mentioned this pull request 3 hours ago
Enable maxfeerate and maxburnamount as startup config options. #29217
Open
Merge state
Add more commits by pushing to the 2023-11-submitpackage-max-fee-burn branch on instagibbs/bitcoin.

All checks have passed
1 skipped and 15 successful checks
@github-actions
CI / test each commit (pull_request) Skipped
Details
@github-actions
CI / macOS 13 native, x86_64, no depends, sqlite only, gui (pull_request) Successful in 34m
Details
@github-actions
CI / Win64 native, VS 2022 (pull_request) Successful in 25m
Details
@cirrus-ci
32-bit CentOS, dash, gui Successful in 21m — Task Summary
Details
@cirrus-ci
ARM, unit tests, no functional tests Successful in 27m — Task Summary
Details
@cirrus-ci
ASan + LSan + UBSan + integer, no depends, USDT Successful in 40m — Task Summary
Details
@cirrus-ci
MSan, depends Successful in 29m — Task Summary
Details
@cirrus-ci
TSan, depends, gui Successful in 32m — Task Summary
Details
@cirrus-ci
Win64, unit tests, no gui tests, no functional tests Successful in 15m — Task Summary
Details
@cirrus-ci
fuzzer,address,undefined,integer, no depends Successful in 42m — Task Summary
Details
@cirrus-ci
lint Successful in 1m — Task Summary
Details
@cirrus-ci
macOS-cross 11.0, gui, no tests Successful in 20m — Task Summary
Details
@cirrus-ci
multiprocess, i686, DEBUG Successful in 36m — Task Summary
Details
@cirrus-ci
no wallet, libbitcoinkernel Successful in 14m — Task Summary
Details
@cirrus-ci
previous releases, qt5 dev package and depends packages, DEBUG Successful in 25m — Task Summary
Details
@cirrus-ci
tidy Successful in 46m — Task Summary
Details
Merging is blocked
The base branch restricts merging to authorized users. Learn more about protected branches.
 You’re not authorized to merge this pull request.
@instagibbs


Add a comment
Comment
 
Add your comment here...
 
Remember, contributions to this repository should follow its contributing guidelines and security policy.
 ProTip! Add .patch or .diff to the end of URLs for Git’s plaintext views.
Reviewers
@glozow
glozow
@ismaelsadeeq
ismaelsadeeq
Still in progress?
Assignees
No one assigned
Labels
RPC/REST/ZMQ
Projects
None yet
Milestone
No milestone
Development
Successfully merging this pull request may close these issues.

 submitpackage: add args "maxfeerate" and "maxburnamount"
5 participants
@instagibbs
@DrahtBot
@[ismaelsadeeq](https://github.com/ismaelsadeeq)
@kashifs
@[glozow](https://github.com/glozow)
Allow edits and access to secrets by maintainers

Reply:
> Do you mean incentive incompatible here? because in the given example if C is rejected (A,B) will be added to the mempool with mining score of 5s/vb which is lower than whats intended.

correct, typo

Reply:
I understand why we should not check the package fee rate here because, in some arrangements, e.g., a parent transaction P will be cfp'ed by some child E, and E can have another child F, which is on its own. 
Checking the package fee rate in this case is incorrect; we should instead evaluate using chunks fee rate later, as you said  👍🏾